zerver/views/__init__.py

```diff
@@ -1048,8 +1050,23 @@ def json_upload_file(request, user_profile):
 @has_request_variables
 def get_uploaded_file(request, realm_id, filename,
                       redir=REQ(validator=check_bool, default=True)):
+    user_profile = request.user
+    path_id = "%s/%s" % (realm_id, filename)
+    # Check if user can view the file
+    is_authorized = validate_attachment_request(user_profile, path_id)
+    if is_authorized == None:
+        return json_error("That file does not exist.", status=404)
+    if not is_authorized:
+        return json_error("You are not authorized to view this file.", status=403)
```

Comment:
I'm not totally sure `json_error` will produce the right result; can you try testing this manually by uploading a file, deleting it on disk, and then reloading the page?